src/lib/updating-element.ts

```diff
@@ -0,0 +1,552 @@
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/**
+ * Converts property values to and from attribute values.
+ */
+interface AttributeSerializer<T = any> {
+
+  /**
+   * Deserializing function called to convert an attribute value to a property value.
+   */
+  fromAttribute?(value: string): T;
+
+  /**
+   * Serializing function called to convert a property value to an attribute value.
+   */
+  toAttribute?(value: T): string|null;
+
+}
+
+type AttributeType<T = any> = AttributeSerializer<T>|((value: string) => T);
+
+/**
+ * Defines options for a property accessor.
+ */
+export interface PropertyDeclaration<T = any> {
+
+  /**
+   * Describes how and whether the property becomes an observed attribute.
+   * If the value is `false`, the property is not added to `observedAttributes`.
+   * If true or absent, the lowercased property name is observed (e.g. `fooBar` becomes `foobar`).
+   * If a string, the string value is observed (e.g `attribute: 'foo-bar'`).
+   */
+  attribute?: boolean|string;
+
+  /**
+   * Describes how to serialize and deserialize the attribute to/from a property.
+   * If this value is a function, it is used to deserialize the attribute value
+   * a the property value. If it's an object, it can have keys for `fromAttribute` and
+   * `toAttribute` where `fromAttribute` is the deserialize function and `toAttribute`
+   * is a serialize function used to set the property to an attribute. If no `toAttribute`
+   * function is provided and `reflect` is set to `true`, the property value is set
+   * directly to the attribute.
+   */
+  type?: AttributeType<T>;
+
+  /**
+   * Describes if the property should reflect to an attribute.
+   * If `true`, when the property is set, the attribute is set using the
+   * attribute name determined according to the rules for the `attribute`
+   * propety option and the value of the property serialized using the rules from
+   * the `type` property option.
+   */
+  reflect?: boolean;
+
+  /**
+   * Describes if setting a property should trigger invalidation and updating.
+   * This function takes the `newValue` and `oldValue` and returns `true` if
+   * invalidation should occur. If not present, a strict identity check
+   * (eg. === operator) is used. This is useful if a property should be
+   * considered dirty only if some condition is met, like if a key of an
+   * object value changes.
+   */
+  shouldInvalidate?(value: T, oldValue: T): boolean;
+
+}
+
+/**
+ * Object that describes accessors to be created on the element prototype.
```

Comment:
super nit: You don't need phrases like "Object that", you can just start with "Describes...". Also, the second sentence doesn't describe the interface, but some other thing that uses it.

Reply:
Rewrote